server/src/main/java/org/candlepin/bind/HandleCertificates.java

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright (c) 2009 - 2012 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.bind;
+
+import org.candlepin.controller.EntitlementCertificateGenerator;
+import org.candlepin.model.Entitlement;
+import org.candlepin.model.EntitlementCertificate;
+import org.candlepin.model.EntitlementCertificateCurator;
+import org.candlepin.model.EntitlementCurator;
+import org.candlepin.model.Pool;
+import org.candlepin.model.PoolQuantity;
+import org.candlepin.model.Product;
+
+import com.google.inject.Inject;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by vrjain on 5/4/17.
+ */
+public class HandleCertificates implements BindOperation {
+
+    private EntitlementCertificateGenerator ecGenerator;
+    private EntitlementCertificateCurator ecCurator;
+    private EntitlementCurator eCurator;
+
+    public HandleCertificates(EntitlementCertificateGenerator ecGenerator, EntitlementCertificateCurator
+        ecCurator, EntitlementCurator eCurator) {
+        this.ecGenerator = ecGenerator;
+        this.ecCurator = ecCurator;
+        this.eCurator = eCurator;
+    }
+
+    private Map<String, EntitlementCertificate> certs;
+
+    private boolean isApplicable = true;
+
+    @Override
+    public void preProcess(BindContext context, BindChain chain) {
+
+        if (context.getConsumer().isShare()) {
+            isApplicable = false;
+            return;
+        }
+
+        Map<String, Product> products = new HashMap<String, Product>();
+        Map<String, PoolQuantity> poolQuantities = context.getPoolQuantities();
+        for (PoolQuantity poolQuantity : poolQuantities.values()) {
+            Pool pool = poolQuantity.getPool();
+            products.put(pool.getId(), pool.getProduct());
+        }
+
+        certs = ecGenerator.generateEntitlementCertificates(context.getConsumer(),
+            products,
+            poolQuantities,
+            context.getEntitlementMap());
+
+        chain.preProcess(context);
```

Comment:
@barnabycourt each command is responsible for calling the next item in the chain. the way it invokes the next item in the chain is by invoking the respective method on the chain itself, which is aware of the next command in line. I believe this design was inspired by the servlet filter interfaces.